great_expectations/datasource/fluent/config.py

```diff
@@ -111,13 +107,25 @@ def _load_datasource_subtype(cls, v: Dict[str, dict]):
                     f"'{ds_name}' has unsupported 'type' - {type_lookup_err}"
                 ) from type_lookup_err
 
-            config["name"] = ds_name
+            if "name" in config:
+                if config["name"] != ds_name:
+                    raise ValueError(
+                        f'Datasource name "{ds_name}" is different from value in its "name" entry.'
```

Comment:
I would show both values.
```suggestion
                    raise ValueError(
                        f'Datasource name "{ds_name}" is different from value in its "name" entry "{<OTHER_VALUE>}".'
```

Reply:
@Kilo59 I incorporated your suggestion for the name change -- for both `Datasource` (key/name) and `DataAsset` key/name as well as your suggestion to report both the key and the name when they do not match (and updated the tests accordingly).  However, I did not create `MISSING` -- we have not used this pattern, and I did not want to create a global sentinel just for this; I feel that the current pattern of test/act (test for key's presence, act if it exists) is easy to read and works well.  Thank you.

Reply:
@alexsherstinsky I think it's okay if you decide not to use this pattern but it is very common in programming.
I would be surprised if we don't do this at least once in the `great_expectations` codebase.

There's even a draft pep right now that does a good job of explaining the need for the pattern. The rest of the pep is not relevant to this conversation.
https://peps.python.org/pep-0661/

![image](https://user-images.githubusercontent.com/13108583/224717021-8f7c664f-5f68-4ff7-8491-d894c4d1fbc4.png)

Reply:
@Kilo59 I have used this pattern in other projects; however, I have not seen it in GreatExpectations (at least not in a widely utilized way).  Introducing it in a truly global way seems like should be part of a separate change (which would need to be discussed with the team).  Thanks.